src/GridFS/GridFsUpload.php

```diff
@@ -0,0 +1,169 @@
+            "filename" => $filename,
+            "uploadDate" => $uploadDate,
+            "_id" => $objectId
+        ];
+
+        $fileOptions = [];
+        if (isset($options['contentType'])) {
+            if (is_string($options['contentType'])) {
+                $fileOptions['contentType'] = $options['contentType'];
+            } else {
+                throw new InvalidArgumentTypeException('"contentType" option', $options['contentType'], 'string');
+            }
+        }
+        if (isset($options['aliases'])) {
+            if (\MongoDB\is_string_array($options['aliases'])) {
+                $fileOptions['aliases'] = $options['aliases'];
+            } else {
+                throw new InvalidArgumentTypeException('"aliases" option', $options['aliases'], 'array of strings');
+            }
+        }
+
+        if (isset($options['metadata'])) {
+            if (is_array($options['metadata']) || is_object($options['metadata'])) {
+                $fileOptions['metadata'] = $options['metadata'];
+            } else {
+                throw new InvalidArgumentTypeException('"metadata" option', $options['metadata'], 'object or array');
+            }
+        }
+        $this->file = array_merge($main_file, $fileOptions);
+        parent::__construct($bucket);
+    }
+    /**
+    * Reads data from a stream into GridFS
+    *
+    * @param Stream   $source   Source Stream
+    * @return ObjectId
+    */
+    public function uploadFromStream($source)
+    {
+        $this->bucket->ensureIndexes();
+
+        if (!is_resource($source) || get_resource_type($source) != "stream") {
+            throw new UnexpectedTypeException('stream', $source);
+        } else{
+            $streamMetadata = stream_get_meta_data($source);
+        } if (!is_readable($streamMetadata['uri'])) {
+     //       throw new InvalidArgumentException("stream not readable");
+            //issue being that php's is_readable reports native streams as not readable like php://temp
+        }
+        while ($data = fread($source, $this->bucket->getChunkSizeBytes())) {
+            $this->insertChunk($data);
+        }
+        return $this->fileCollectionInsert();
+    }
+    /**
+    * Insert a chunks into GridFS from a string
+    *
+    * @param string  $toWrite   String to upload
+    * @return int
+    */
+    public function insertChunks($toWrite)
+    {
+        $this->bucket->ensureIndexes();
+
+        $readBytes = 0;
+        while($readBytes != strlen($toWrite)) {
+            $addToBuffer = substr($toWrite, $readBytes, $this->bucket->getChunkSizeBytes() - $this->bufferLength);
+            fwrite($this->buffer, $addToBuffer);
+            $readBytes += strlen($addToBuffer);
+            $this->bufferLength += strlen($addToBuffer);
+            if($this->bufferLength == $this->bucket->getChunkSizeBytes()) {
+                rewind($this->buffer);
+                $this->insertChunk(fread($this->buffer, $this->bucket->getChunkSizeBytes()));
```

Comment:
It may be easier to just use [`stream_get_contents()`](http://php.net/stream_get_contents) on the buffer here, as it shouldn't contain more than a chunk's worth of bytes.